### PR TITLE
adding structured patterns, adding ranking

### PR DESCRIPTION
**Pull request by cmungall**
- Added slot rank, for ordering of elements. Proposed model for structured patterns, addresses https://github.com/linkml/linkml/issues/176
- illustrates https://github.com/linkml/linkml/issues/176
- expanding examples

### CODE REVIEW

File: examples/pattern-example.yaml [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+#==================================
+# Slots                           #
+#==================================
+
+slots:
+  id:
+    identifier: true
+    range: string
+    pattern: "^P\\d{7}"    ## simple pattern. Note {}s are NOT interpolated
```
Comment (ddooley) [inline examples/pattern-example.yaml:56]: I see now that the simple "pattern" allows flexibility in start or end of string matching.  Meanwhile as you have it, the structured_pattern, involving string components, can't have re's that have start or end markers.

File: examples/pattern-example.yaml [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+  name:
+    range: string
+  email:
+    range: EmailString
+  height:
+    range: string
+    structured_pattern:
+      pattern: "{float} {unit.length}"
```
Comment (ddooley) [inline examples/pattern-example.yaml]: Technical question, can a structured_pattern pattern have a mixture of {setting} and re expressions, like  "{float}\s*{unit.length}" ?

Comment (ddooley) [reply examples/pattern-example.yaml]: Side note, it seems like for javascript based programming, at runtime load of a linkml spec, each given "pattern" can be converted into a compiled regular expression by expanding it out and then re.compile() ?  I've done this for very efficient runtime validation of data.

Comment (cmungall) [reply examples/pattern-example.yaml]: Q1: I had envisioned yes, and that most implementations would interpolate directly to a regex.

JS: yes, that's the idea. And stay tuned re JS and linkml...

**Pull request merged**